src/Adapter/Product/Update/CombinationStockUpdater.php

```diff
@@ -0,0 +1,144 @@
+use PrestaShop\PrestaShop\Core\Domain\Product\Combination\Exception\CannotUpdateCombinationException;
+use PrestaShop\PrestaShop\Core\Domain\Product\Combination\Exception\CombinationConstraintException;
+use PrestaShop\PrestaShop\Core\Domain\Product\Combination\ValueObject\CombinationId;
+use PrestaShop\PrestaShop\Core\Domain\Product\Stock\Exception\StockAvailableNotFoundException;
+use PrestaShop\PrestaShop\Core\Exception\CoreException;
+use PrestaShop\PrestaShop\Core\Stock\StockManager;
+use PrestaShopException;
+
+class CombinationStockUpdater
+{
+    /**
+     * @var StockAvailableRepository
+     */
+    private $stockAvailableRepository;
+
+    /**
+     * @var CombinationRepository
+     */
+    private $combinationRepository;
+
+    /**
+     * @var StockManager
+     */
+    private $stockManager;
+
+    /**
+     * @param StockAvailableRepository $stockAvailableRepository
+     * @param CombinationRepository $combinationRepository
+     * @param StockManager $stockManager
+     */
+    public function __construct(
+        StockAvailableRepository $stockAvailableRepository,
+        CombinationRepository $combinationRepository,
+        StockManager $stockManager
+    ) {
+        $this->stockAvailableRepository = $stockAvailableRepository;
+        $this->combinationRepository = $combinationRepository;
+        $this->stockManager = $stockManager;
+    }
+
+    /**
+     * @param Combination $combination
+     * @param array $propertiesToUpdate
+     * @param bool $addMovement
+     *
+     * @throws CombinationConstraintException
+     * @throws CoreException
+     * @throws StockAvailableNotFoundException
+     */
+    public function update(Combination $combination, array $propertiesToUpdate, bool $addMovement): void
+    {
+        $this->combinationRepository->partialUpdate(
+            $combination,
+            $propertiesToUpdate,
+            CannotUpdateCombinationException::FAILED_UPDATE_STOCK
+        );
+
+        // updating stockAvailable after Combination is important, because quantity validation is done in Combination update
+        if (in_array('quantity', $propertiesToUpdate)) {
+            $this->updateStockAvailable($combination, $addMovement);
+        }
+    }
+
+    /**
+     * @param Combination $combination
+     * @param bool $addMovement
+     *
+     * @throws CombinationConstraintException
+     * @throws CoreException
+     * @throws StockAvailableNotFoundException
+     */
+    private function updateStockAvailable(Combination $combination, bool $addMovement): void
+    {
+        $newQuantity = (int) $combination->quantity;
+        $stockAvailable = $this->stockAvailableRepository->getForCombination(new CombinationId((int) $combination->id));
+
+        if ($addMovement) {
+            $this->saveMovement($combination, (int) $stockAvailable->quantity, $newQuantity);
+        }
+
+        $stockAvailable->quantity = $newQuantity;
```

Comment:
location must be updated in StockAvailable as well

Reply:
This actually makes me realize something, but StockAvailable is the one that contains the `out_of_stock` setting but it seems only the one attached to the Product is important, we can't edit it by combination

Reply:
so this means now Combination cannot run out of stock? :thinking: